sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMemoryLeaseManager.scala

```diff
@@ -0,0 +1,406 @@
+      return new MemoryLease {
+        override def leaseAmount: Long = 0
+
+        override def close(): Unit = {}
+      }
+    }
+
+    val totalAmountRequested = requiredAmount + optionalAmount
+    val amountToRequest = if (nonBlocking && totalAmountRequested > memoryAvailable) {
+      logDebug(s"Task: $taskAttemptId requested $totalAmountRequested bytes in a " +
+          s"non-blocking request, but only $memoryAvailable is available. The request will " +
+          s"be reduced to $memoryAvailable.")
+      memoryAvailable
+    } else if (totalAmountRequested + alreadyRequested > memoryForLease) {
+      logWarning(s"Task: $taskAttemptId requested $totalAmountRequested bytes, but has " +
+          s"already requested $alreadyRequested bytes. This would go over the total for the " +
+          s"worker $memoryForLease reducing the request on the hope that this was an " +
+          s"overestimate.")
+      memoryForLease - alreadyRequested
+    } else {
+      totalAmountRequested
+    }
+
+    val task = activeTasks.computeIfAbsent(taskAttemptId, _ => new TrackingForTask)
+
+    if (!taskPriorityQueue.isEmpty || memoryAvailable < amountToRequest) {
+      // Need to block until there is memory available
+      if (task.isThreadWaiting || task.canWakeUp) {
+        throw new IllegalStateException("MULTIPLE THREADS TRYING TO REQUEST RESOURCES FOR " +
+            "A SINGLE TASK")
+      }
+      // We should never hit this, because the checks before should avoid blocking, but just to be
+      // safe we are going to check.
+      assert(!nonBlocking)
+
+      // Release all of the currently used memory so it can be reused by another task if needed.
+      memoryAvailable += task.totalUsed
+      wakeTasksThatCanRun()
+
+      // Update the task request state for what is now needed
+      task.isThreadWaiting = true
+      task.totalUsed += amountToRequest
+      taskPriorityQueue.offer(task)
+      withResource(new NvtxRange("Acquire GPU Memory", NvtxColor.YELLOW)) { _ =>
+        while (!task.canWakeUp) {
+          try {
+            wait()
+          } catch {
+            case _: InterruptedException =>
+          }
+          if (shuttingDown) {
+            throw new IllegalStateException("GpuMemoryLeaseManager shut down")
+          }
+        }
+      }
+
+      // Memory available was already updated when the task was marked for wakeup.
+      task.canWakeUp = false
+      task.isThreadWaiting = false
+    } else {
+      // No need to block we can just do this...
+      memoryAvailable -= amountToRequest
+      task.totalUsed += amountToRequest
+    }
+    new MemoryLease {
+      override def leaseAmount: Long = amountToRequest
+
+      override def close(): Unit = {
+        releaseLease(tc, amountToRequest)
+      }
+    }
+  }
+
+  def getTotalLease(tc: TaskContext): Long = synchronized {
+    val taskAttemptId = tc.taskAttemptId()
+    val data = activeTasks.get(taskAttemptId)
```

Comment:
nit, instead of `data` here and below in `releaseLease` we could use `task` as we did above.